dEdX/src/dEdxAnalyser.cc

```diff
@@ -571,19 +698,60 @@ void dEdxAnalyser::processEvent( LCEvent * evt ) {
         _MCmomVec.push_back(mcpar0->getMomentum()[1]);
         _MCmomVec.push_back(mcpar0->getMomentum()[2]);
       }
-    }
 
-    // check which hits are inside the TPC
-    TrackerHitVec trkHits = track->getTrackerHits();
-    int nHitsTPC = 0;
-    for(unsigned int iHit=0; iHit<trkHits.size(); iHit++){
-      if(trkHits[iHit]->getType() == 0)
+      // check which hits are inside the TPC
+      TrackerHitVec trkHits = track->getTrackerHits();
+      int nHitsTPC = 0;
+      for(unsigned int iHit=0; iHit<trkHits.size(); iHit++)
       {
-        double x = trkHits[iHit]->getPosition()[0];
-        double y = trkHits[iHit]->getPosition()[1];
-        if(sqrt(x*x+y*y)>=_TPC_inner && sqrt(x*x+y*y)<=_TPC_outer) nHitsTPC++;
+        if(trkHits[iHit]->getType() == 0)
+        {
+          double x = trkHits[iHit]->getPosition()[0];
+          double y = trkHits[iHit]->getPosition()[1];
+          if(sqrt(x*x+y*y)>=_TPC_inner && sqrt(x*x+y*y)<=_TPC_outer) nHitsTPC++;
+        }
       }
-    }
+
+      // if track hits are available check hit spectrum
+      //if(_useHitCol)
+      //{
+        //const TrackerHitVec Hits = track->getTrackerHits();
+        for (unsigned int h=0; h<trkHits.size(); ++h)
+        {
+          double EDep = trkHits[h]->getEDep();
+          for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HitEnergySpectrum[j]->Fill(EDep);
+          if (pdg==10) _HitEnergySpectrum[_nPart]->Fill(EDep);
+          _HitEnergySpectrum[_nPart+1]->Fill(EDep);
+          if (.9<mom && mom<1.15)  // around 1 GeV
+          {
+            for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HitEnergySpectrum_at1[j]->Fill(EDep);
+            if (pdg==10) _HitEnergySpectrum_at1[_nPart]->Fill(EDep);
+            _HitEnergySpectrum_at1[_nPart+1]->Fill(EDep);
+          }
+          if (2<mom && mom<3)  // around 2.5 GeV
+          {
+            for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HitEnergySpectrum_at25[j]->Fill(EDep);
+            if (pdg==10) _HitEnergySpectrum_at25[_nPart]->Fill(EDep);
+            _HitEnergySpectrum_at25[_nPart+1]->Fill(EDep);
+          }
+          if (1) //(mom_t>=1 && nHits>=200)
+          {
+            for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HEHist[j]->Fill(mom,EDep);
+            if (pdg==10) _HEHist[_nPart]->Fill(mom,EDep);
+            _HEHist[_nPart+1]->Fill(mom,EDep);
+            //for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HEHist_bg[j]->Fill(mom/_Masses[j],EDep);
+            //for (int j=0; j<_nPart; j++) if (pdg==_PDG[j]) _HEHist_bg[_nPart+1]->Fill(mom/_Masses[j],EDep);
+          }
+        }
```

Comment:
Can this be restructured? It looks like it should be possible to find the index of the histogram that has to be filled only once and then use this index throughout the loop of the tracker hits, instead of finding it again for every histogram.